### PR TITLE
Steps towards a11y for Redwood Router

### PR DESCRIPTION
**Pull request by jtoar**
> At first this PR was just for scroll, but I started adding all the other a11y changes as well so I could test them all together. More than happy to split this up when the time comes.

The goal of this PR is to make Redwood Router as accessible as possible. All of the work done here is based on the work done at Gatbsy, in particular Marcy Sutton's [research](https://www.gatsbyjs.com/blog/2019-07-11-user-testing-accessible-client-routing) and Madalyn Parker's PRs (namely this one: https://github.com/gatsbyjs/gatsby/pull/26376). This wouldn't be possible without the work they did. 🙏 

In a nutshell, this PR tries to make so that the router automatically manages scroll, announcements, and focus, and adds four components for when the user wants to do so specifically:

- RouteAnnouncement (initial implementation's one-to-one with Gatsby's)
- RouteFocus (initial implementation's one-to-one with Gatsby's)
- SkipNavLink (Reach UI)
- SkipNavContent (Reach UI)

While the motivation for most of the changes is feature-parity, there's plenty of room to add Redwood-specific additions like: 
- page and layout generation with RouteAnnouncements, RouteFocus, SkipNavLinks, SkipNavContents
- automatic insertion of SkipNavLinks, SkipNavContents where it makes sense to do so (i.e. layouts)&mdash;this in particular would be a great feature to have

## scroll

`pushState` scroll in Redwood is broken (here's a  <a href="https://s.tape.sh/OEU0XBIQ">Tape</a> showing it), and it's actually pretty remarkable that there's only been one, quite issue about it (https://github.com/redwoodjs/redwood/issues/1362). This PR provides a fix for it. Here's another <a href="https://s.tape.sh/0iWcB3m5">Tape</a> showing the behavior with this PR. You should see that the page being navigated to is scrolled to the top, even with [Fast](https://s.tape.sh/B8wxKi0R) and [Slow 3G](https://s.tape.sh/im9mcIkY) on.

> The hard part about this&mdash;and this goes for most of this PR&mdash;wasn't really what line to add (it was always gonna be `global?.scrollTo`, etc.), but where to add it.

The expected behavior is that, on clicking a link (which is just a call to `navigate`), you're scrolled to the top of the page you're navigating to. The actual behavior is that we're left exactly where we were instead. 

But note that while `pushState` scroll is broken, `popstate` isn't. `popstate` is the event that fires when the user hits the back or forward button. Redwood has one `popstate` listener; it updates the location provider, which renders the entire app. So it's kind of important.

The expected scroll behavior for `popstate` is that scroll's restored&mdash;put the user back where they were. And that's what happens. Here's another <a href="https://s.tape.sh/ssYHVCxA">Tape</a> showing it (I'm navigating backwards through the same set of pages, Contact -> About -> Home).

## announcements

With SPAs, there's never a full-page refresh (that's the point after all), but while we gain the more seamless experience that comes with never refreshing the page, we lose things we "had for free" like announcements (i.e. screen readers knew that the page had changed and would say so). Unless we manage them ourselves, we risk making the web a less accessible place.

On page load, the router tried to announce something as specific as it can by looking for announcements in this order (in a prerender-friendly way!):

1. RouteAnnouncement
2. h1
3. document.title
4. location.pathname

If there's a RouteAnnouncement, that means the user provided it, so it's as specific as it gets. And if the user provided many, it looks for the most specific one. 

> Note that specificity != descriptive, which is to say that doesn't mean it's good&mdash;there needs to be validation to ensure that a hand-made announcement is still a good one (i.e. not too long&mdash;think a git commit message).

The previous behavior was announcing nothing and never knowing that the page had changed, so even if RouteAnnouncement's aren't at-large in Redwood Apps, this is still a huge improvement.

## focus

This is the most work-in-progress section. Currently, on navigation, focus is left where it was. In tandem with announcing that the route has changed, focus should be set to somewhere that makes sense&mdash;either the top of the DOM or to the new content&mdash;to something that the user can quickly interact with in a meaningful way. Marcy's research suggests that that interactive element be a skip link. This is the behavior we're aiming for.

## notes and todos

- [x] add flags to the page / layout generators to include RouteAnnouncements and RouteFocus + SkipNavLink and SkipNavContent
  - e.g. `yarn rw g page home / --a11y`, or `yarn rw g layout main --a11y`
- [ ] is it ok to use Reach UI?
  - figured at first why reinvent the wheel, plus it's a standard. One thing to do still if it is ok:
  - [ ] export appropriate Reach UI CSS so that the skip link's invisible until tabbed
- [x] tests & CI
- [ ] corresponding docs at https://github.com/redwoodjs/redwoodjs.com/pull/567
- [x] type things properly
- [ ] put all this in `@redwoodjs/router/a11y`, like `@redwoodjs/web/toast`?

### CODE REVIEW

File: packages/router/src/page-loader.tsx
Comment (dac09) [inline packages/router/src/page-loader.tsx:105]: Careful, careful :) Prerender will not like this change!

```suggestion
    typeof window !== 'undefined' && window.scrollTo(0, 0)
```

Comment (Tobbe) [inline packages/router/src/page-loader.tsx:105]:
```suggestion
    window?.scrollTo(0, 0)
```

Comment (dac09) [reply packages/router/src/page-loader.tsx:105]: That won't work I'm afraid! It'll still throw

Comment (Tobbe) [reply packages/router/src/page-loader.tsx:105]: @dac09 You've got your work cut out for you now, don't you, scanning for stuff like this in every PR we do 😅

Comment (dac09) [reply packages/router/src/page-loader.tsx:105]: Not looking forward to it! But anything `router` in particular, we have to watch out for!

Comment (dac09) [reply packages/router/src/page-loader.tsx:105]: Here's another way actually (please do verify it works)
```suggestion
    global?.scrollTo(0, 0)
```

Comment (Tobbe) [reply packages/router/src/page-loader.tsx:105]: Why/how? I thought `window` would just be `undefined`

Comment (dac09) [reply packages/router/src/page-loader.tsx:105]: Try this in your browser console. 

`console.log(bazinga?.notThere)`

It throws because the safe access operator needs atleast the object to be defined.

Comment (Tobbe) [reply packages/router/src/page-loader.tsx:105]: I see. I thought `window` would actually be `=== undefined`, so `console.log(undefined?.notThere)`. But now I understand that it's not even available at all.

Comment (dac09) [reply packages/router/src/page-loader.tsx:105]: Hmmm... this is a good point, maybe I can shim window to be undefined in the prerender process, so that we can use this clean syntax!

Comment (jtoar) [reply packages/router/src/page-loader.tsx:105]: this line, `global?.scrollTo(0, 0)`, works @dac09

Comment (jtoar) [reply packages/router/src/page-loader.tsx:105]: @dac09 Is this the way that you think we'll go with?

Comment (dac09) [reply packages/router/src/page-loader.tsx:105]: I think I'd rather have a linter rule to be honest, because "monkey patching" window, even if its to undefined - makes me a little worried

**Comment by jtoar** [comment]
Most of the comments will obviously be removed; just using them to save some of the ideas I had while reviewing with others.

**Comment by peterp** [comment]
Thanks @jtoar, just read through this, looking forward to our discussion a bit later

File: packages/router/src/util.ts
Comment (peterp) [inline packages/router/src/util.ts:264]: I suggestion to make this feel a bit more robust is to "return early", so find and validate the element your looking for, and return if it matches:


```js
export const getAnnouncement = () => {
   // find route annoucements element
  const ra = globals?.document.querySelectorAll( '[data-redwood-route-announcement]')?.[0]
  
  if (ra?.textContent) {
    return ra.textContent
  }
  
  /* ... */
```

**Review by peterp — APPROVED** [review]
Nice one! I've left a suggestion about the utility function, but otherwise it looks good.

Please wait for Tobbe to merge his Set PR before merging this (in case of conflicts).

**Comment by Tobbe** [comment]
> Please wait for Tobbe to merge his Set PR before merging this (in case of conflicts).

Done!

@jtoar Let me know if I can help in any way with merges etc

File: packages/router/src/__tests__/set.test.tsx
Comment (jtoar) [inline packages/router/src/__tests__/set.test.tsx:70]: @Tobbe I had to update `set.test.tsx` to include the markup for the route announcer (rendered by the `PageLoader`)--double checking that this doesn't screw things up for you?

Comment (Tobbe) [reply packages/router/src/__tests__/set.test.tsx:70]: Yeah, I kind of wish <Set> was less tied to the router, but I had to add the stuff that checks if it has any matching children or not. So changes to the router unfortunately affects the Set component and tests as well.

Your changes look 👍

Comment (jtoar) [reply packages/router/src/__tests__/set.test.tsx:70]: Makes sense, and I will work on decoupling the announcement in the future, don't like it either! All the announcement really needs to know is when the page is rendered, and right now the only way to really know that is in `componentDidUpdate`

**Comment by Tobbe** [comment]
Just noticed this when looking at your code

image

That's on me, and I'll get a PR up to fix it!

**Comment by jtoar** [comment]
@Tobbe nice! ok to merge this in the mean time?

**Comment by Tobbe** [comment]
This is great Dom 👍 
Merge at will!

**Pull request merged**
